// omnisharp-client/lib/enums.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.isPluginDriver = isPluginDriver;
var Driver = exports.Driver = undefined;
(function (Driver) {
    Driver[Driver["Http"] = 0] = "Http";
    Driver[Driver["Stdio"] = 1] = "Stdio";
})(Driver || (exports.Driver = Driver = {}));
var DriverState = exports.DriverState = undefined;
(function (DriverState) {
    DriverState[DriverState["Disconnected"] = 0] = "Disconnected";
    DriverState[DriverState["Downloading"] = 1] = "Downloading";
    DriverState[DriverState["Downloaded"] = 2] = "Downloaded";
    DriverState[DriverState["Connecting"] = 3] = "Connecting";
    DriverState[DriverState["Connected"] = 4] = "Connected";
    DriverState[DriverState["Error"] = 5] = "Error";
})(DriverState || (exports.DriverState = DriverState = {}));
var Runtime = exports.Runtime = undefined;
(function (Runtime) {
    Runtime[Runtime["ClrOrMono"] = 0] = "ClrOrMono";
    Runtime[Runtime["CoreClr"] = 1] = "CoreClr";
})(Runtime || (exports.Runtime = Runtime = {}));
function isPluginDriver(driver) {
    return !!driver.updatePlugins;
}
