// omnisharp-client/languageserver/server-extended.js
"use strict";
var Methods;
(function (Methods) {
    var Extended;
    (function (Extended) {
        Extended.GetCodeActionsRequest = '__extended/textDocument/getCodeActions';
        Extended.RunCodeActionRequest = '__extended/textDocument/runCodeAction';
        Extended.ImplementationRequest = '__extended/textDocument/implementation';
        Extended.NavigateRequest = '__extended/textDocument/navigate';
        Extended.PublishHighlightNotification = '__extended/textDocument/publishHighlight';
    })(Extended = Methods.Extended || (Methods.Extended = {}));
})(Methods = exports.Methods || (exports.Methods = {}));
/**
 * A request to rename a symbol.
 */
var GetCodeActionsRequest;
(function (GetCodeActionsRequest) {
    GetCodeActionsRequest.type = { get method() { return Methods.Extended.GetCodeActionsRequest; } };
})(GetCodeActionsRequest = exports.GetCodeActionsRequest || (exports.GetCodeActionsRequest = {}));
/**
 * A request to rename a symbol.
 */
var RunCodeActionRequest;
(function (RunCodeActionRequest) {
    RunCodeActionRequest.type = { get method() { return Methods.Extended.RunCodeActionRequest; } };
})(RunCodeActionRequest = exports.RunCodeActionRequest || (exports.RunCodeActionRequest = {}));
/**
 * A request to find implementation
 */
var ImplementationRequest;
(function (ImplementationRequest) {
    ImplementationRequest.type = { get method() { return Methods.Extended.ImplementationRequest; } };
})(ImplementationRequest = exports.ImplementationRequest || (exports.ImplementationRequest = {}));
/**
 * A request to find implementation
 */
var NavigateRequest;
(function (NavigateRequest) {
    NavigateRequest.type = { get method() { return Methods.Extended.NavigateRequest; } };
})(NavigateRequest = exports.NavigateRequest || (exports.NavigateRequest = {}));
/**
 * Diagnostics notification are sent from the server to the client to signal
 * results of validation runs.
 */
var HighlightNotification;
(function (HighlightNotification) {
    HighlightNotification.type = { get method() { return Methods.Extended.PublishHighlightNotification; } };
})(HighlightNotification = exports.HighlightNotification || (exports.HighlightNotification = {}));
